http-clients/src/main/java/feign/TraceResponseDecoder.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ */
+
+package feign;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Iterables;
+import com.palantir.tracing.TraceState;
+import com.palantir.tracing.Traces;
+import feign.codec.DecodeException;
+import feign.codec.Decoder;
+import java.io.IOException;
+import java.lang.reflect.Type;
+
+public final class TraceResponseDecoder implements Decoder {
+
+    private final Decoder delegate;
+
+    public TraceResponseDecoder(Decoder delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Object decode(Response response, Type type) throws IOException, DecodeException, FeignException {
+        String traceId = safeGetOnlyElement(response.headers().get(Traces.Headers.TRACE_ID), null);
+        String spanId = safeGetOnlyElement(response.headers().get(Traces.Headers.SPAN_ID), null);
+        Optional<TraceState> trace = Traces.getTrace();
+        if (traceId != null && spanId != null && trace.isPresent()) {
```

Comment:
we could consider always popping the top of the trace here, in theory every single request should be matched by a decode call.